cmd/snap-preseed/preseed_linux.go

```diff
@@ -200,7 +207,188 @@ func chooseTargetSnapdVersion() (*targetSnapdInfo, error) {
+		{"--bind", underWritable("system-data/var/cache/snapd"), underPreseed("var/cache/snapd")},
+		{"--bind", underWritable("system-data/var/cache/apparmor"), underPreseed("var/cache/apparmor")},
+		{"--bind", underWritable("system-data/var/snap"), underPreseed("var/snap")},
+		{"--bind", underWritable("system-data/snap"), underPreseed("snap")},
+		{"--bind", underWritable("system-data/etc/systemd"), underPreseed("etc/systemd")},
+		{"--bind", underWritable("system-data/etc/dbus-1"), underPreseed("etc/dbus-1")},
+		{"--bind", underWritable("system-data/etc/udev/rules.d"), underPreseed("etc/udev/rules.d")},
+		{"--bind", filepath.Join(snapdMountPath, "/usr/lib/snapd"), underPreseed("/usr/lib/snapd")},
+		{"--bind", filepath.Join(prepareImageDir, "system-seed"), underPreseed("var/lib/snapd/seed")},
+	}
+
+	for _, mountArgs := range mounts {
+		cmd := exec.Command("mount", mountArgs...)
+		if out, err = cmd.CombinedOutput(); err != nil {
+			return nil, fmt.Errorf("cannot prepare mountpoint in preseed mode: %v\n'mount %s' failed with: %s", err, strings.Join(mountArgs, " "), out)
+		}
+		mounted = append(mounted, mountArgs[len(mountArgs)-1])
+	}
+
+	return cleanupMounts, nil
+}
+
+func systemForPreseeding(systemsDir string) (label string, err error) {
+	systemLabels, err := filepath.Glob(filepath.Join(systemsDir, "systems", "*"))
+	if err != nil && !os.IsNotExist(err) {
+		return "", fmt.Errorf("cannot list available systems: %v", err)
+	}
+	if len(systemLabels) != 1 {
+		return "", fmt.Errorf("expected a single system for preseeding, found %d", len(systemLabels))
+	}
+	return filepath.Base(systemLabels[0]), nil
+}
+
+var makePreseedTempDir = func() (string, error) {
+	return ioutil.TempDir("", "preseed-")
+}
+
+var makeWritableTempDir = func() (string, error) {
+	return ioutil.TempDir("", "writable-")
+}
+
+func prepareCore20Chroot(prepareImageDir string) (preseed *PreseedOpts, cleanup func(), err error) {
+	sysDir := filepath.Join(prepareImageDir, "system-seed")
+	sysLabel, err := systemForPreseeding(sysDir)
+	snapdSnapPath, baseSnapPath, err := systemSnapFromSeed(sysDir, sysLabel)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if snapdSnapPath == "" {
+		return nil, nil, fmt.Errorf("snapd snap not found")
+	}
+	if baseSnapPath == "" {
+		return nil, nil, fmt.Errorf("base snap not found")
+	}
+
+	tmpPreseedDir, err := makePreseedTempDir()
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot prepare uc20 chroot: %v", err)
+	}
+	writableTmpDir, err := makeWritableTempDir()
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot prepare uc20 chroot: %v", err)
+	}
+
+	cleanupMounts, err := prepareCore20Mountpoints(prepareImageDir, tmpPreseedDir, snapdSnapPath, baseSnapPath, writableTmpDir)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot prepare uc20 mountpoints: %v", err)
+	}
+
+	cleanup = func() {
+		cleanupMounts()
+		if err := os.RemoveAll(tmpPreseedDir); err != nil {
+			fmt.Fprintf(Stdout, "%v", err)
```

Comment:
is there a reason not to use Printf() here?

Reply:
Yes, Stdout can be mocked in the tests of this package (although this particular error case isn't tested).

Reply:
Ah! OK :-)